helix-rest/src/main/java/org/apache/helix/rest/metadatastore/TrieRoutingData.java

```diff
@@ -0,0 +1,136 @@
+    TrieNode leafNode = findTrieNode(path, true);
+    return leafNode._realmAddress;
+  }
+
+  /**
+   * If findLeafAlongPath is false, then starting from the root node, find the trie node that the
+   * given path is pointing to and return it; raise IllegalArgumentException if the path does
+   * not point to any node. If findLeafAlongPath is true, then starting from the root node, find the
+   * leaf node along the provided path; raise IllegalArgumentException if the path does not
+   * point to any node or if there is no leaf node along the path.
+   * @param path - the path where the search is conducted
+   * @param findLeafAlongPath - whether the search is for a leaf node on the path
+   * @return the node pointed by the path or a leaf node along the path
+   * @throws IllegalArgumentException - when the path points to nothing or when no leaf node is found
+   */
+  private TrieNode findTrieNode(String path, boolean findLeafAlongPath)
+      throws IllegalArgumentException {
+    if (path.equals("/") || path.equals("")) {
+      if (findLeafAlongPath && !_rootNode._isLeaf) {
+        throw new IllegalArgumentException("no leaf node found along the path");
+      }
+      return _rootNode;
+    }
+
+    if (path.substring(0, 1).equals("/")) {
+      path = path.substring(1);
+    }
+    String[] splitPath = path.split("/", 0);
+
+    TrieNode curNode = _rootNode;
+    if (findLeafAlongPath && curNode._isLeaf) {
+      return curNode;
+    }
+    Map<String, TrieNode> curChildren = curNode._children;
+    for (String pathSection : splitPath) {
+      curNode = curChildren.get(pathSection);
+      if (curNode == null) {
+        throw new IllegalArgumentException("the provided path is missing from the trie");
+      }
+      if (findLeafAlongPath && curNode._isLeaf) {
+        return curNode;
+      }
+      curChildren = curNode._children;
+    }
+    if (findLeafAlongPath) {
+      throw new IllegalArgumentException("no leaf node found along the path");
+    }
+    return curNode;
+  }
+
+  /**
+   * Given a trie node, search for all leaf nodes that descend from the given trie node, and add
+   * all complete paths and realm addresses of these leaf nodes to a provided mapping. This method
+   * recursively calls itself.
+   * @param mapping - where the results (complete paths of leaf nodes and their realm addresses)
+   *          are stored
+   * @param curNode - the current trie node where the search starts
+   * @param curPath - the complete path of the current trie node
+   */
+  private static void addAllAddressesToMapping(Map<String, String> mapping, TrieNode curNode,
+      String curPath) {
+    if (curNode._isLeaf) {
+      mapping.put(curPath, curNode._realmAddress);
+      return;
+    }
+
+    for (Map.Entry<String, TrieNode> entry : curNode._children.entrySet()) {
+      addAllAddressesToMapping(mapping, entry.getValue(), curPath + "/" + entry.getKey());
```

Comment:
What if the path does not exist? Why we dont use a complete path to add to trie but with partial info trie node to add to the path? Logically, it is not correct.

Reply:
Two points are raised here which I'll try to break down: 
1. What if the path does not exist?
The path has to exist because it's traversing the previously constructed trie; whatever path is defined will be in the trie - this method is not creating arbitrary paths. I believe logically this method is correct. I have several unit tests to prove that it's working.
2. Why don't we store the complete path in the trie node?
I have considered this design before, and I believe storing complete paths also work. However, partial paths is working without an issue, therefore I see no significant benefit changing it. If you feel strongly about this, I'm happy to discuss further.

Reply:
Update on point 2 (also replied on the previous comment): the "complete path" is marginally better for an iterative `getAllMappingUnderPath` therefore I added the full path field to `TrieNode`. Thanks for the suggestion @dasahcc . Leaving this conversation unresolved for point 1, point 2 has been resolved.